src/sync/sys/windows/net.rs

```diff
@@ -0,0 +1,303 @@
+        })
+    }
+
+    pub(crate) fn accept(
+        &self,
+        quit_flag: &Arc<AtomicBool>,
+    ) -> std::result::Result<Option<PipeConnection>, io::Error> {
+        if quit_flag.load(Ordering::SeqCst) {
+            info!("listener shutdown for quit flag");
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                "listener shutdown for quit flag",
+            ));
+        }
+
+        // Create a new pipe for every new client
+        let mut namedpipe = self.new_instance().unwrap();
+
+        let mut poll = Poll::new().unwrap();
+        let mut events = Events::with_capacity(1024);
+        poll.registry()
+            .register(
+                &mut namedpipe,
+                SERVER,
+                Interest::WRITABLE,
+            )
+            .unwrap();
+
+        loop {
+            match namedpipe.connect() {
+                Ok(()) => {
+                    // pipe is locked so can't use it here.
+                    break;
+                }
+                Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => {
+                    println!("waiting for client to connect");
+                    poll.poll(&mut events, None).unwrap();
+                    continue;
+                }
+                Err(e) => {
+                    println!("Error connecting to pipe: {e}");
+                    return Err(e);
+                }
+            }
+        }
+
+        poll.registry()
+            .reregister(
+                &mut namedpipe,
+                SERVER,
+                Interest::READABLE,
+            )
+            .unwrap();
+        let instance_num = self.instance_number.fetch_add(1, Ordering::SeqCst);
+        trace!("pipe instance {} connected", instance_num);
+        let pipe_instance = PipeConnection {
+            named_pipe: Mutex::new(namedpipe),
+            poller: Mutex::new(poll),
+            instance_number: instance_num,
+        };
+        Ok(Some(pipe_instance))
+    }
+
+    fn new_instance(&self) -> io::Result<NamedPipe> {
+        let name = OsStr::new(&self.address.as_str())
+            .encode_wide()
+            .chain(Some(0)) // add NULL termination
+            .collect::<Vec<_>>();
+
+        // bitwise or file_flag_first_pipe_instance with file_flag_overlapped and pipe_access_duplex
+        let mut openmode = PIPE_ACCESS_DUPLEX | FILE_FLAG_OVERLAPPED;
+
+        if self.first_instance.load(Ordering::SeqCst) {
+            openmode |= FILE_FLAG_FIRST_PIPE_INSTANCE;
+            self.first_instance.swap(false, Ordering::SeqCst);
+        }
+
+        // Safety: syscall
+        let h = unsafe {
+            CreateNamedPipeW(
+                name.as_ptr(),
+                openmode,
+                PIPE_TYPE_BYTE,
+                PIPE_UNLIMITED_INSTANCES,
+                65536,
+                65536,
+                0,
+                std::ptr::null_mut(), // todo set this on first instance
+            )
+        };
```

Comment:
We should add `PIPE_REJECT_REMOTE_CLIENTS`

Reply:
added

Reply:
@dcantah I was looking into the security descriptor. A null value sets the default security descriptor which only allows Admins and the current user (https://learn.microsoft.com/en-us/windows/win32/ipc/named-pipe-security-and-access-rights).  

I was comparing to the way [go-wino creates](https://github.com/microsoft/go-winio/blob/650a2e464c7458a1d355d7bb5affad1eec1bfbb6/pipe.go#L291-L304) it, And I believe this is the same behavior although go-winio is using undocumented calls (`ntCreateNamedPipeFile` and `rtlDefaultNpAcl`) so has the additional calls to `rtlDefaultNpAcl` to set the default. In the case of the documented `CreateNamedPipeW` null provides a sane default of requiring admins.